feat: table sorting II

deboer-tim opened the pull request:
### What does this PR do?

Fixes the sort issues raised in issue #4881:

- Fixed sort direction: ascending sort should be 'up', descending 'down'.
- Added a new 'initialSortDir' for Columns like Size to specify that they should be sorted in descending order by default. This could have been a boolean flag, but seeing "initialSortDir: 'descending'" in Column definitions seems more natural and obvious to me.

This also fixes the sorting gap raised in #4860: constant sorting by any column.

Prior to this change sorting only happens once, when the user clicks on the column header. If the data changes it becomes unsorted again, and the column header is then wrong/out of sync. e.g. if you're in the Containers view, sort by Status, and then start a new container, data will revert back to the initial order and the new container will appear where it would have, even though the Status column still says it is sorted.

Now, sorting is also applied _whenever the data changes_. This means that if you sort by Status and start a container it would move to the top, or if you've sorted Age by descending it would move to the bottom. i.e. the sorting is always applied. This also means that any view can set the Table defaultSortColumn to any arbitrary column.

### Screenshot/screencast of this PR

Note changed sort indicators (ascending is 'up'), Containers sort descending the first time you click on it (initialSortDir), and once sorted by Status the started pod moves to the top: 
https://github.com/containers/podman-desktop/assets/19958075/bc3c69ee-150b-44a5-b1eb-d48575a64d47

### What issues does this PR fix or reference?

Fixes #4881.

### How to test this PR?

Use along with the Images or Volumes PR.

benoitf commented on packages/renderer/src/lib/table/Table.svelte:
```diff
@@ -40,6 +40,10 @@ function toggleAll(checked: boolean) {
 let sortCol: Column<any>;
 let sortAscending: boolean;
 
+$: if (data) {
+  sortImpl();
+}
+
```
I would move the condition to the sort but it's only my feeling

deboer-tim replied:
I would too, but it is using svelte reactivity here: sortImpl() is called whenever data changes to make sure the sort is maintained. e.g. if we get an update from provider that replaces all the objects the current sort is still applied.

benoitf replied:
I mean, to do:

```
$: sortImpl();
```

and in sortImpl if we don't have data, we abort

deboer-tim replied:
Yes, but if you remove `if (data)` then sortImpl() never gets called. Including data in the `$:` statement triggers svelte to register data as reactive and call sortImpl() whenever it gets changed.

benoitf commented on packages/renderer/src/lib/table/table.ts:
```diff
@@ -47,6 +47,19 @@ export interface ColumnInformation<Type> {
+   *
+   * Defaults to 'ascending'.
+   */
+  readonly initialSortDir?: 'ascending' | 'descending';
```
could be just be named `order` ?  (I mean, of course user can change it later but when we register objects we specify an order) (or then `initialOrder`

deboer-tim replied:
Renamed to initialOrder in new commit.

benoitf reviewed (approved):
I guess it requires a test ?

benoitf commented on packages/renderer/src/lib/table/Table.svelte:
```diff
@@ -55,8 +59,21 @@ function sort(column: Column<any>) {
   } else {
     sortCol = column;
-    sortAscending = true;
+    sortAscending = column.info.initialSortDir ? column.info.initialSortDir !== 'descending' : true;
```
```suggestion
    sortAscending = column.info.initialOrder ? column.info.initialOrder !== 'descending' : true;
```

deboer-tim commented:
Rebased since the changes couldn't be auto-merged on main due to other PRs. While I was at it I added a couple brackets for consistent indenting, and added the sortOrder to the Volume page Size column so that there's actually something to test. Clicking on Size (once you've pulled them) will sort by largest first and now set the column sort indicator correctly (down).

I take back the comment I made in the description about this change _always_ keeping data sorted, as I've found a case in the Volumes page where data changes and yet there is no svelte trigger. I'll investigate separately, but this change is still at least a partial change in that direction: it wasn't the primary focus of this PR but more pages will stay sorted when data changes after this PR.

deboer-tim commented:
Additional test added.
